File: app/src/main/java/nl/acidcats/tumblrlikes/ui/widgets/PhotoNavBar.java
package nl.acidcats.tumblrlikes.ui.widgets;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.LocalBroadcastManager;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import nl.acidcats.tumblrlikes.R;
import nl.acidcats.tumblrlikes.data.constants.Broadcasts;
import nl.acidcats.tumblrlikes.data.constants.FilterType;
import nl.acidcats.tumblrlikes.ui.widgets.filterdropdown.FilterDropdown;
import nl.acidcats.tumblrlikes.ui.widgets.filterdropdown.FilterOptionSelectionListener;

/**
 * Created by stephan on 16/05/2017.
 */

public class PhotoNavBar extends FrameLayout {
    private static final String TAG = PhotoNavBar.class.getSimpleName();

    @BindView(R.id.btn_filter)
    TextView _filterButton;
    @BindView(R.id.btn_settings)
    View _settingsButton;
    @BindView(R.id.btn_refresh)
    View _refreshButton;
    @BindView(R.id.filter_dropdown)
    FilterDropdown _filterDropdown;

    private Unbinder _unbinder;
    private FilterOptionSelectionListener _filterOptionSelectionListener;

    public PhotoNavBar(@NonNull Context context) {
        super(context);

        init();
    }

    public PhotoNavBar(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        init();
    }

    public PhotoNavBar(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init();
    }

    private void init() {
        View view = LayoutInflater.from(getContext()).inflate(R.layout.navbar, this, true);
        _unbinder = ButterKnife.bind(this, view);

        _filterButton.setOnClickListener(this::onFilterButtonClick);
        _settingsButton.setOnClickListener(this::onSettingsButtonClick);
        _refreshButton.setOnClickListener(this::onRefreshButtonClick);

        _filterButton.setText(R.string.filter_all);

        _filterDropdown.setFilterOptionSelectionListener(filterType -> setFilterType(filterType, true));

        hide();
    }

    public void setFilterType(FilterType filterType) {
        setFilterType(filterType, false);
    }

    private void setFilterType(FilterType filterType, boolean notifyListener) {
        if (_filterOptionSelectionListener != null) {
            _filterOptionSelectionListener.onOptionSelected(filterType);
        }

        _filterDropdown.hide();

        _filterButton.setText(filterType.getResId());
    }

    public void setFilterOptionSelectionListener(FilterOptionSelectionListener listener) {
        _filterOptionSelectionListener = listener;
    }

    public void show() {
        setVisibility(VISIBLE);
    }

    public void hide() {
        setVisibility(GONE);

        _filterDropdown.hide();
    }

    private void onRefreshButtonClick(View view) {
        sendBroadcast(Broadcasts.REFRESH_REQUEST);
    }

    private void onSettingsButtonClick(View view) {
        sendBroadcast(Broadcasts.SETTINGS_REQUEST);
    }

    private void sendBroadcast(String action) {
        LocalBroadcastManager.getInstance(getContext()).sendBroadcast(new Intent(action));
    }

    private void onFilterButtonClick(View view) {
        _filterDropdown.show();
    }

    public void onDestroy() {
        _filterButton.setOnClickListener(null);
        _settingsButton.setOnClickListener(null);
        _refreshButton.setOnClickListener(null);

        _unbinder.unbind();
    }
}

File: app/src/main/java/nl/acidcats/tumblrlikes/ui/screens/photo_screen/widgets/PhotoNavBar.kt
package nl.acidcats.tumblrlikes.ui.screens.photo_screen.widgets

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import com.github.ajalt.timberkt.Timber
import kotlinx.android.synthetic.main.navbar.view.*
import nl.acidcats.tumblrlikes.R
import nl.acidcats.tumblrlikes.core.constants.FilterType
import nl.acidcats.tumblrlikes.ui.screens.photo_screen.PhotoScreenContract

/**
 * Created on 31/10/2018.
 */
class PhotoNavBar @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {

    var filterTypeSelectedListener: FilterTypeSelectedListener? = null
    var navBarListener: PhotoScreenContract.NavBarListener? = null

    init {
        LayoutInflater.from(context).inflate(R.layout.navbar, this, true)

        filterButton.setOnClickListener { filterDropdown.show() }
        settingsButton.setOnClickListener { navBarListener?.onSettingsRequested() }
        refreshButton.setOnClickListener { navBarListener?.onRefreshRequested() }

        filterDropdown.filterTypeSelectedListener = { setFilter(it, true) }

        visibility = View.GONE
    }

    fun show() {
        visibility = View.VISIBLE

        if (height == 0) {

        }

        translationY = -height.toFloat()
        animate()
                .setDuration(resources.getInteger(android.R.integer.config_shortAnimTime).toLong())
                .translationY(0f)
                .setListener(null)
    }

    fun hide() {
        Timber.d { "hide: " }

        filterDropdown.hide()

        animate()
                .setDuration(resources.getInteger(android.R.integer.config_shortAnimTime).toLong())
                .translationY(-height.toFloat())
                .setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator?) {
                        visibility = View.GONE
                    }
                })
    }

    fun setFilter(filterType: FilterType) = setFilter(filterType, false)

    private fun setFilter(filterType: FilterType, notifyListener: Boolean) {
        filterButton.text = filterDropdown.getFilterLabel(filterType)

        filterDropdown.hide()

        if (notifyListener) {
            filterTypeSelectedListener?.invoke(filterType)
        }
    }

    fun enableRefreshButton(enabled: Boolean) {
        Timber.d { "enableRefreshButton: $enabled" }

        refreshButton.isEnabled = enabled
        refreshButton.alpha = if (enabled) 1f else .2f
    }

    fun onDestroyView() {
        filterButton.setOnClickListener(null)
        settingsButton.setOnClickListener(null)
        refreshButton.setOnClickListener(null)
    }
}
